gk/fib.c

```diff
@@ -0,0 +1,1337 @@
+				neigh_fib->action != GK_FWD_NEIGHBOR_BACK_NET &&
+				neigh_fib->action != GK_FWD_GATEWAY_BACK_NET)) {
+			return NULL;
+		}
+
+		/* The gateway FIB entry exists. */
+		if (neigh_fib->action == action)
+			return neigh_fib;
+
+		/* Create a new FIB entry for the gateway. */
+		gw_fib = add_prefix_fib(&gw_addr,
+			(sizeof(struct in_addr) * 8), gk_conf);
+		if (gw_fib == NULL)
+			return NULL;
+
+		eth_cache = neigh_add_ether_cache(
+			&neigh_fib->u.neigh, &gw_addr);
+		if (eth_cache == NULL) {
+			rte_lpm_delete(ltbl->lpm,
+				ntohl(gw_addr.ip.v4.s_addr),
+				(sizeof(struct in_addr) * 8));
+			return NULL;
+		}
+	} else if (likely(gw_addr.proto == ETHER_TYPE_IPv6)
+			&& ipv6_if_configured(iface)) {
+		fib_id = lpm_lookup_ipv6(ltbl->lpm6, gw_addr.ip.v6.s6_addr);
+		/*
+		 * Invalid gateway entry, since at least we should
+		 * obtain the FIB entry for the neighbor table.
+		 */
+		if (fib_id < 0)
+			return NULL;
+
+		neigh_fib = &ltbl->fib_tbl6[fib_id];
+
+		/*
+		 * Invalid gateway entry, since the neighbor entry
+		 * and the gateway entry should be in the same network.
+		 */
+		if ((action == GK_FWD_GATEWAY_FRONT_NET &&
+				neigh_fib->action != GK_FWD_NEIGHBOR_FRONT_NET
+				&& neigh_fib->action !=
+				GK_FWD_GATEWAY_FRONT_NET)
+				|| (action == GK_FWD_GATEWAY_BACK_NET &&
+				neigh_fib->action != GK_FWD_NEIGHBOR_BACK_NET &&
+				neigh_fib->action != GK_FWD_GATEWAY_BACK_NET)) {
+			return NULL;
+		}
+
+		/* The gateway FIB entry exists. */
+		if (neigh_fib->action == action)
+			return neigh_fib;
+
+		/* Create a new FIB entry for the gateway. */
+		gw_fib = add_prefix_fib(&gw_addr,
+			(sizeof(struct in6_addr) * 8), gk_conf);
+		if (gw_fib == NULL)
+			return NULL;
+
+		eth_cache = neigh_add_ether_cache(
+			&neigh_fib->u.neigh6, &gw_addr);
+		if (eth_cache == NULL) {
+			rte_lpm6_delete(ltbl->lpm6,
+				gw_addr.ip.v6.s6_addr,
+				(sizeof(struct in6_addr) * 8));
+			return NULL;
+		}
+	} else
+		rte_panic("Unexpected condiction: invalid IP type %hu at interface %s!\n",
+			gw_addr.proto, iface->name);
+
+	/* Fill up the fib entry for the gateway. */
+	gw_fib->action = action;
+
+	rte_memcpy(&gw_fib->u.gateway.ip_addr,
+		&gw_addr, sizeof(gw_fib->u.gateway.ip_addr));
+
+	gw_fib->u.gateway.eth_cache = eth_cache;
+	eth_cache->stale = true;
+	eth_cache->eth_hdr.ether_type = gw_addr.proto;
+
+	if (action == GK_FWD_GATEWAY_FRONT_NET) {
+		ether_addr_copy(&gk_conf->net->front.eth_addr,
+			&eth_cache->eth_hdr.s_addr);
+	} else {
+		ether_addr_copy(&gk_conf->net->back.eth_addr,
+			&eth_cache->eth_hdr.s_addr);
+	}
+
+	eth_cache->ref_cnt++;
+
+	/*
+	 * Add a fib entry for the IP prefix,
+	 * for which the gateway is responsible.
+	 *
+	 * Notice, for this fib, it doesn't need to initialize the
+	 * fields @flow and @eth_cache, since it's only used to help
+	 * lookup the fib entry for the gateway.
+	 */
+	ip_prefix_fib = add_prefix_fib(
+		&ip_prefix_addr, ip_prefix_len, gk_conf);
+	if (ip_prefix_fib != NULL) {
```

Comment:
Although not all call to `add_prefix_fib()` are wrong, I think that rewriting the code to avoid it even during the initialization would make the code uniform and avoid the need for `add_prefix_fib()`; at least in its current format. The advantage of dropping `add_prefix_fib()` is that the concurrency bug would not sneak into the code due to an eventual, accidental misuse of `add_prefix_fib()`.